### PR TITLE
TPC attach 2 framework

### PR DESCRIPTION
**Pull request by ofirgo**
## Pull Request Description:

Implementing TPC Attach2Fw module to attach a given TPC model to the framework-specific operators.
The attachments is based on a hardcoded mapping between the TPC schema's OperatorSetNames enum to each framework layers and ops (keras/pytorch).

Note that current TPCs do not use this module, since they will be removed and replaced with a simpler initialization framework in the future. So currently, this module is not used in MCT but it is just a preparation for future modifications.

## Checklist before requesting a review:
- [ ] I set the appropriate labels on the pull request.
- [ ] I have added/updated the release note draft (if necessary).
- [ ] I have updated the documentation to reflect my changes (if necessary).
- [ ] All function and files are well documented. 
- [ ] All function and classes have type hints.
- [ ] There is a licenses in all file.
- [ ] The function and variable names are informative. 
- [ ] I have checked for code duplications.
- [ ] I have added new unittest (if necessary).

### CODE REVIEW

File: model_compression_toolkit/target_platform_capabilities/schema/v1.py [hunk 1]
```diff
@@ -61,7 +61,6 @@ class OperatorSetNames(Enum):
     OPSET_DROPOUT = "Dropout"
     OPSET_SPLIT = "Split"
     OPSET_CHUNK = "Chunk"
-    OPSET_UNBIND = "Unbind"
```
Comment (ofirgo) [inline model_compression_toolkit/target_platform_capabilities/schema/v1.py:64]: @lior-dikstein ignore this, since you already remove this in your PR

File: model_compression_toolkit/target_platform_capabilities/target_platform/targetplatform2framework/attach2keras.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+            OperatorSetNames.OPSET_FAKE_QUANT_WITH_MIN_MAX_VARS.value: [tf.quantization.fake_quant_with_min_max_vars],
+            OperatorSetNames.OPSET_COMBINED_NON_MAX_SUPPRESSION.value: [tf.image.combined_non_max_suppression],
+            OperatorSetNames.OPSET_CROPPING2D.value: [Cropping2D],
+            OperatorSetNames.OPSET_ZERO_PADDING2d.value: [ZeroPadding2D],
+            OperatorSetNames.OPSET_CAST.value: [tf.cast],
+            OperatorSetNames.OPSET_STRIDED_SLICE.value: [tf.strided_slice]
+        }
+
+        if FOUND_SONY_CUSTOM_LAYERS:
+            self._opset2layer[OperatorSetNames.OPSET_POST_PROCESS] = [SSDPostProcess]
```
Comment (ofirgo) [inline model_compression_toolkit/target_platform_capabilities/target_platform/targetplatform2framework/attach2keras.py:94]: @lior-dikstein assuming this is the operator set name that is going to be added to the enum for pre process

File: model_compression_toolkit/target_platform_capabilities/target_platform/targetplatform2framework/attach2keras.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+            OperatorSetNames.OPSET_BATCH_NORM.value: [BatchNormalization],
+            OperatorSetNames.OPSET_RELU.value: [tf.nn.relu, ReLU],
+            OperatorSetNames.OPSET_RELU6.value: [tf.nn.relu6],
+            OperatorSetNames.OPSET_LEAKY_RELU.value: [tf.nn.leaky_relu, LeakyReLU],
+            OperatorSetNames.OPSET_HARD_TANH.value: [LayerFilterParams(Activation, activation="hard_tanh")],
+            OperatorSetNames.OPSET_ADD.value: [tf.add, Add],
+            OperatorSetNames.OPSET_SUB.value: [tf.subtract, Subtract],
+            OperatorSetNames.OPSET_MUL.value: [tf.math.multiply, Multiply],
+            OperatorSetNames.OPSET_DIV.value: [tf.math.divide, tf.math.truediv],
+            OperatorSetNames.OPSET_MIN_MAX.value: [tf.math.minimum, tf.math.maximum, Minimum, Maximum],
```
Comment (haihabi) [inline model_compression_toolkit/target_platform_capabilities/target_platform/targetplatform2framework/attach2keras.py:65]: Split two operator  min & max

File: model_compression_toolkit/target_platform_capabilities/target_platform/targetplatform2framework/attach2fw.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        # that has weights attributes with provided quantization config (in the tp model) to
+        # its framework-specific attribute name.
+        # note that a DefaultDict should be provided if not all the layer types in the
+        # operation set are provided separately in the mapping.
+        self._opset2attr_mapping = None
+
+    def attach(self, tpc_model: TargetPlatformModel,
+               custom_opset2layer: Dict[str, Tuple[List[Any], Optional[Dict[str, DefaultDict]]]] = None
+               ) -> TargetPlatformCapabilities:
+
```
Comment (lior-dikstein) [inline model_compression_toolkit/target_platform_capabilities/target_platform/targetplatform2framework/attach2fw.py:24]: Need to add documentation

File: model_compression_toolkit/target_platform_capabilities/target_platform/targetplatform2framework/attach2fw.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+from model_compression_toolkit.target_platform_capabilities.schema.mct_current_schema import TargetPlatformModel
+from model_compression_toolkit.target_platform_capabilities.target_platform import TargetPlatformCapabilities, \
+    OperationsSetToLayers
+
+
+class TpcAttach2Fw:
+
+    def __init__(self):
+        self._opset2layer = None
+
```
Comment (lior-dikstein) [inline model_compression_toolkit/target_platform_capabilities/target_platform/targetplatform2framework/attach2fw.py:13]: why not get it as an input and eliminate the context manager functionality? then you also document it with type hint and documentation

File: model_compression_toolkit/target_platform_capabilities/target_platform/targetplatform2framework/attach2fw.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+        self._opset2attr_mapping = None
+
+    def attach(self, tpc_model: TargetPlatformModel,
+               custom_opset2layer: Dict[str, Tuple[List[Any], Optional[Dict[str, DefaultDict]]]] = None
+               ) -> TargetPlatformCapabilities:
+
+        tpc = TargetPlatformCapabilities(tpc_model)
+
+        with tpc:
+            for opset_name, operators in self._opset2layer.items():
```
[the remark this answers is not in the document]
Comment (ofirgo) [reply model_compression_toolkit/target_platform_capabilities/target_platform/targetplatform2framework/attach2fw.py:28]: I don't mind keeping it for now in the TPC initialization because it is not part of the schema and I prefer as minimal unessential changes that can break things

File: model_compression_toolkit/target_platform_capabilities/target_platform/targetplatform2framework/attach2keras.py (hunk 3 above)
Comment (ofirgo) [reply model_compression_toolkit/target_platform_capabilities/target_platform/targetplatform2framework/attach2keras.py:65]: @lior-dikstein this is more relevant to your PR

File: model_compression_toolkit/target_platform_capabilities/target_platform/targetplatform2framework/attach2fw.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
+from typing import Dict, Tuple, List, Any, Optional
+
+from model_compression_toolkit import DefaultDict
+from model_compression_toolkit.target_platform_capabilities.schema.mct_current_schema import TargetPlatformModel
+from model_compression_toolkit.target_platform_capabilities.target_platform import TargetPlatformCapabilities, \
+    OperationsSetToLayers
+
+
+class TpcAttach2Fw:
```
[the remark this answers is not in the document]
Comment (ofirgo) [reply model_compression_toolkit/target_platform_capabilities/target_platform/targetplatform2framework/attach2fw.py:9]: I wanted it to begin with TPC to put it into context, I think it's clearer, but I wouldn't mind changing it

File: model_compression_toolkit/target_platform_capabilities/target_platform/targetplatform2framework/attach2fw.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+        self._opset2layer = None
+
+        # we provide attributes mapping that maps each layer type in the operations set
+        # that has weights attributes with provided quantization config (in the tp model) to
+        # its framework-specific attribute name.
+        # note that a DefaultDict should be provided if not all the layer types in the
+        # operation set are provided separately in the mapping.
+        self._opset2attr_mapping = None
+
+    def attach(self, tpc_model: TargetPlatformModel,
```
[the remark this answers is not in the document]
Comment (ofirgo) [reply model_compression_toolkit/target_platform_capabilities/target_platform/targetplatform2framework/attach2fw.py:21]: this is debatable, but @haihabi want us to refer to what currently known as "TP model", meaning the platform description that is going to be given as a JSON, as "TPC model".
so I'm starting to change the terminology inside the MCT

File: model_compression_toolkit/target_platform_capabilities/target_platform/targetplatform2framework/attach2fw.py (hunk 5 above)
Comment (ofirgo) [reply model_compression_toolkit/target_platform_capabilities/target_platform/targetplatform2framework/attach2fw.py:13]: As I said, I prefer not to deal with the TPC context manager elimination as part of this refactor

File: model_compression_toolkit/target_platform_capabilities/target_platform/targetplatform2framework/attach2fw.py (hunk 4 above)
Comment (lior-dikstein) [reply model_compression_toolkit/target_platform_capabilities/target_platform/targetplatform2framework/attach2fw.py:36]: """
    Attach a TargetPlatformModel, which describes platform capabilities, to specific framework operators.
    
    Args:
        tpc_model: A TargetPlatformModel object.
        custom_opset2layer: An optional dictionary to add or override the built-in framework operator sets. 
            This dictionary maps a unique operator set name to a pair consisting of:
            1. A list of framework operators.
            2. An optional mapping of operator attribute names.

    Returns:
        A TargetPlatformCapabilities object.
    """

**Review by lior-dikstein — APPROVED** [review]
